File: runners/ruby/negative_v1.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# Ruby runner for the rfc9421_negative_v1 CORE cross-language battery.
#
# Independent Ruby port of the verdict surface (signing_base.py, parse.py,
# keycheck.py, verify.py + the algovoi-rfc9421-ecdsa provider). Self-contained
# (no Ruby RFC 9421 verifier exists); must produce byte-identical verdicts to
# every other runner. Ed25519 verify via OpenSSL (key built from a hand-assembled
# SPKI DER, since OpenSSL 3.0 here has no new_raw_public_key), with an explicit
# canonical-S (S < L) malleability check and the ported small-order / non-canonical
# key gate applied fail-closed first. ECDSA P-256/P-384 via OpenSSL with explicit
# range / width / on-curve / strict-low-s checks.
#
# Corpus path: ARGV[0], else $ALGOVOI_NEGATIVE_V1, else the sibling repo default.

require "json"
require "base64"
require "openssl"

DEFAULT_PATH = "../../corpus/rfc9421_negative_v1/rfc9421_negative_v1.json"

class SigningBaseError < StandardError; end
class ParseError < StandardError; end
class WeakKeyError < StandardError; end

# ================= signing base (port of signing_base.py) =================
def build_signing_base(input, mode, sig_params_raw)
  raise SigningBaseError, "bad mode" unless %w[algovoi-v0 rfc9421].include?(mode)
  raise SigningBaseError, "rfc9421 mode requires signature_params_raw" if mode == "rfc9421" && sig_params_raw.nil?

  headers = {}
  (input["headers"] || {}).each { |k, v| headers[k.downcase] = v }
  params = input["parameters"] || {}

  str_or_throw = lambda do |field, msg|
    v = input[field]
    raise SigningBaseError, msg if v.nil?
    v
  end
  param_str_or_throw = lambda do |key, msg|
    raise SigningBaseError, msg unless params.key?(key)
    params[key].to_s
  end

  lines = []
  input["covered_components"].each do |component|
    c = component.downcase
    value =
      case c
      when "@method"
        v = str_or_throw.call("method", "@method covered but method not supplied")
        mode == "rfc9421" ? v : v.downcase
      when "@authority"
        str_or_throw.call("authority", "@authority covered but authority not supplied").downcase
      when "@path"
        str_or_throw.call("path", "@path covered but path not supplied")
      when "@target-uri"
        str_or_throw.call("target_uri", "@target-uri covered but target_uri not supplied")
      when "@scheme"
        str_or_throw.call("scheme", "@scheme covered but scheme not supplied").downcase
      when "@status"
        raise SigningBaseError, "@status covered but status not supplied" if input["status"].nil?
        input["status"].to_s
      when "created"
        param_str_or_throw.call("created", "'created' covered but no 'created' parameter in Signature-Input")
      when "expires"
        param_str_or_throw.call("expires", "'expires' covered but no 'expires' parameter in Signature-Input")
      else
        raise SigningBaseError, "unsupported derived component: #{component}" if c.start_with?("@")
        raise SigningBaseError, "covered header #{component} not present" unless headers.key?(c)
        headers[c]
      end
    lines << "\"#{c}\": #{value}"
  end
  lines << "\"@signature-params\": #{sig_params_raw}" if mode == "rfc9421"
  lines.join("\n")
end

# ================= structured-field parse (port of parse.py) =================
LABEL_RE = /\A\s*([A-Za-z][A-Za-z0-9_-]*)\s*=\s*/
COVERED_RE = /\A\(\s*(?:"[^"]*"\s*)*\)/

def parse_signature_input(header)
  raise ParseError, "header must be str" if header.nil?
  hv = header.strip
  raise ParseError, "empty header value" if hv.empty?
  if (m = LABEL_RE.match(hv))
    rest = hv[m.end(0)..]
  elsif hv.start_with?("(")
    rest = hv
  else
    raise ParseError, "no label or covered-components list found at start"
  end
  raise ParseError, "no covered-components list found" unless COVERED_RE.match?(rest)
end

def parse_signature_value(header)
  raise ParseError, "header must be str" if header.nil?
  hv = header.strip
  raise ParseError, "empty Signature header value" if hv.empty?
  if (m = LABEL_RE.match(hv))
    rest = hv[m.end(0)..].strip
  elsif hv.start_with?(":")
    rest = hv
  else
    raise ParseError, "no label or signature-value prefix found at start"
  end
  raise ParseError, "signature value must be wrapped in colons" unless rest.start_with?(":") && rest.end_with?(":") && rest.length >= 2
  sig_b64 = rest[1..-2]
  begin
    decoded = Base64.strict_decode64(sig_b64)
  rescue ArgumentError
    raise ParseError, "signature value is not valid base64"
  end
  raise ParseError, "non-canonical base64" unless Base64.strict_encode64(decoded) == sig_b64
end

# ================= Ed25519 key gate (port of keycheck.py) =================
P = 2**255 - 19
D = (-121665 * 121666.pow(P - 2, P)) % P
SQRT_M1 = 2.pow((P - 1) / 4, P)
L_ORDER = 2**252 + 27_742_317_777_372_353_535_851_937_790_883_648_493

def recover_x(y, sign)
  return nil if y >= P
  y2 = (y * y) % P
  x2 = ((y2 - 1) * (D * y2 + 1).pow(P - 2, P)) % P
  return (sign == 1 ? nil : 0) if x2.zero?
  x = x2.pow((P + 3) / 8, P)
  x = (x * SQRT_M1) % P unless ((x * x - x2) % P).zero?
  return nil unless ((x * x - x2) % P).zero?
  x = P - x if (x & 1) != sign
  x
end

def decode_point(pk)
  return nil unless pk.bytesize == 32
  full = pk.reverse.unpack1("H*").to_i(16)
  sign = (full >> 255) & 1
  y = full & ((1 << 255) - 1)
  x = recover_x(y, sign)
  return nil if x.nil?
  [x, y, 1, (x * y) % P]
end

def point_add(p, q)
  a = ((p[1] - p[0]) * (q[1] - q[0])) % P
  b = ((p[1] + p[0]) * (q[1] + q[0])) % P
  c = (2 * p[3] * q[3] * D) % P
  dd = (2 * p[2] * q[2]) % P
  e = b - a; f = dd - c; g = dd + c; h = b + a
  [(e * f) % P, (g * h) % P, (f * g) % P, (e * h) % P]
end

def mul8(p)
  p2 = point_add(p, p); p4 = point_add(p2, p2); point_add(p4, p4)
end

def identity?(p)
  (p[0] % P).zero? && ((p[1] - p[2]) % P).zero?
end

def small_order?(pk)
  p = decode_point(pk)
  return false if p.nil?
  identity?(mul8(p))
end

def check_ed25519_public_key(pk)
  p = decode_point(pk)
  raise WeakKeyError, "non-canonical/off-curve" if p.nil?
  raise WeakKeyError, "small-order" if identity?(mul8(p))
end

# ================= Ed25519 verify (port of verify.py) =================
def ed25519_verify(msg, sig, pk)
  return false unless sig.bytesize == 64
  return false unless pk.bytesize == 32
  begin
    check_ed25519_public_key(pk)
  rescue WeakKeyError
    return false
  end
  # canonical-S: reject S >= L (Ed25519 signature malleability)
  s = sig[32, 32].reverse.unpack1("H*").to_i(16)
  return false if s >= L_ORDER
  begin
    der = ["302a300506032b6570032100"].pack("H*") + pk
    key = OpenSSL::PKey.read(der)
    key.verify(nil, sig, msg)
  rescue StandardError
    false
  end
end

# ================= ECDSA verify (port of algovoi-rfc9421-ecdsa) =================
def ecdsa_verify(curve, msg, sig_raw, pub, strict_low_s)
  field_size = curve == "p256" ? 32 : 48
  group_name = curve == "p256" ? "prime256v1" : "secp384r1"
  digest = curve == "p256" ? "SHA256" : "SHA384"
  return false unless sig_raw.bytesize == 2 * field_size

  r = sig_raw[0, field_size].unpack1("H*").to_i(16)
  s = sig_raw[field_size, field_size].unpack1("H*").to_i(16)
  group = OpenSSL::PKey::EC::Group.new(group_name)
  n = group.order.to_i
  return false unless r.between?(1, n - 1) && s.between?(1, n - 1)
  return false if strict_low_s && s > n / 2
  return false unless pub.bytesize == 1 + 2 * field_size && pub.bytes[0] == 0x04

  begin
    # Point.new validates the point is on the curve (raises otherwise)
    OpenSSL::PKey::EC::Point.new(group, OpenSSL::BN.new(pub.unpack1("H*"), 16))
    spki = OpenSSL::ASN1::Sequence([
      OpenSSL::ASN1::Sequence([
        OpenSSL::ASN1::ObjectId("id-ecPublicKey"),
        OpenSSL::ASN1::ObjectId(group_name)
      ]),
      OpenSSL::ASN1::BitString(pub)
    ])
    key = OpenSSL::PKey.read(spki.to_der)
    der = OpenSSL::ASN1::Sequence([OpenSSL::ASN1::Integer(r), OpenSSL::ASN1::Integer(s)]).to_der
    key.verify(OpenSSL::Digest.new(digest), der, msg)
  rescue StandardError
    false
  end
end

# ================= driver =================
def hexb(s) = [s].pack("H*")

path = ARGV[0] || ENV["ALGOVOI_NEGATIVE_V1"] || DEFAULT_PATH
corpus = JSON.parse(File.read(path))
total = 0
matched = 0
fails = []

record = lambda do |ok, section, c|
  total += 1
  if ok
    matched += 1
  else
    fails << "[#{section}] #{c['note']}"
  end
end

corpus["signing_base"].each do |c|
  want = Base64.decode64(c["signing_base_b64"]).force_encoding("UTF-8")
  ok =
    begin
      c["ok"] && build_signing_base(c["in"], c["mode"], c["signature_params_raw"]) == want
    rescue StandardError
      !c["ok"]
    end
  record.call(ok, "signing_base", c)
end
corpus["signature_input_parse"].each do |c|
  parsed_ok = begin; parse_signature_input(c["header"]); true; rescue ParseError; false; end
  record.call(parsed_ok == c["ok"], "sig_input_parse", c)
end
corpus["signature_value_parse"].each do |c|
  parsed_ok = begin; parse_signature_value(c["header"]); true; rescue ParseError; false; end
  record.call(parsed_ok == c["ok"], "sig_value_parse", c)
end
corpus["keygate"].each do |c|
  raw = hexb(c["pk_hex"])
  rejected = begin; check_ed25519_public_key(raw); nil; rescue WeakKeyError; "WeakKeyError"; end
  record.call(rejected == c["rejected"] && small_order?(raw) == c["small_order"], "keygate", c)
end
corpus["ed25519_verify"].each do |c|
  base = Base64.decode64(c["signing_base_b64"])
  valid = ed25519_verify(base, hexb(c["sig_hex"]), hexb(c["pk_hex"]))
  record.call(valid == c["expect_valid"], "ed25519_verify", c)
end
corpus["ecdsa_verify"].each do |c|
  valid = ecdsa_verify(c["curve"], hexb(c["msg_hex"]), hexb(c["sig_raw_hex"]),
                       hexb(c["pub_uncompressed_hex"]), c["strict_low_s"] == true)
  record.call(valid == c["expect_valid"], "ecdsa_verify", c)
end

fails.each { |f| puts "FAIL  #{f}" }
puts "\nruby: #{matched}/#{total} cases matched"
exit(matched == total ? 0 : 1)

File: runners/scala/negative_v1.scala
// Scala runner for the rfc9421_negative_v1 / _v2 CORE cross-language battery.
//
// Independent Scala/JVM port of the verdict surface (signing_base.py, parse.py,
// keycheck.py, verify.py + the algovoi-rfc9421-ecdsa provider). Self-contained
// (no Scala RFC 9421 verifier exists); must produce byte-identical verdicts to
// every other runner. Ed25519 verify + canonical-S via Bouncy Castle; the
// small-order / non-canonical key gate is ported with java.math.BigInteger
// (RFC 8032 Appendix A arithmetic); ECDSA P-256/P-384 via the JDK provider with
// explicit range / width / on-curve / strict-low-s checks. JSON via Jackson.
//
// Run: scala-cli run negative_v1.scala -- <corpus path>
// Corpus path: argv[0], else $ALGOVOI_NEGATIVE_V1, else the sibling repo default.

//> using scala "3.8.4"
//> using dep "org.bouncycastle:bcprov-jdk18on:1.78.1"
//> using dep "com.fasterxml.jackson.core:jackson-databind:2.17.0"

import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import org.bouncycastle.asn1.{ASN1EncodableVector, ASN1Integer, DERSequence}
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer

import java.io.File
import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.security.{AlgorithmParameters, KeyFactory, Signature}
import java.security.spec.{ECFieldFp, ECGenParameterSpec, ECParameterSpec, ECPoint, ECPublicKeySpec}
import java.security.spec.{MGF1ParameterSpec, PSSParameterSpec, X509EncodedKeySpec}
import java.util.Base64
import scala.jdk.CollectionConverters.*

class SigningBaseError(m: String) extends RuntimeException(m)
class ParseError(m: String) extends RuntimeException(m)
class WeakKeyError(m: String) extends RuntimeException(m)

object NegativeV1:
  val DefaultPath = "../../corpus/rfc9421_negative_v2/rfc9421_negative_v2.json"

  // ================= signing base (port of signing_base.py) =================
  def buildSigningBase(in: JsonNode, mode: String, spr: String): String =
    if mode != "algovoi-v0" && mode != "rfc9421" then throw SigningBaseError("bad mode")
    if mode == "rfc9421" && spr == null then throw SigningBaseError("rfc9421 needs spr")
    val headers = scala.collection.mutable.LinkedHashMap[String, String]()
    val h = in.get("headers")
    if h != null && h.isObject then
      h.fieldNames().asScala.foreach(k => headers(k.toLowerCase) = h.get(k).asText())
    val params = in.get("parameters")

    def strOr(field: String, msg: String): String =
      val n = in.get(field); if n == null || n.isNull then throw SigningBaseError(msg) else n.asText()
    def paramOr(key: String, msg: String): String =
      if params == null || !params.has(key) then throw SigningBaseError(msg)
      val v = params.get(key); if v.isIntegralNumber then v.bigIntegerValue().toString else v.asText()

    val lines = scala.collection.mutable.ArrayBuffer[String]()
    in.get("covered_components").elements().asScala.foreach { compNode =>
      val component = compNode.asText()
      val c = component.toLowerCase
      val value = c match
        case "@method"     => val v = strOr("method", "@method covered but not supplied"); if mode == "rfc9421" then v else v.toLowerCase
        case "@authority"  => strOr("authority", "@authority covered but not supplied").toLowerCase
        case "@path"       => strOr("path", "@path covered but not supplied")
        case "@target-uri" => strOr("target_uri", "@target-uri covered but not supplied")
        case "@scheme"     => strOr("scheme", "@scheme covered but not supplied").toLowerCase
        case "@status" =>
          val n = in.get("status"); if n == null || n.isNull then throw SigningBaseError("@status covered but not supplied")
          n.bigIntegerValue().toString
        case "created" => paramOr("created", "'created' covered but no created parameter")
        case "expires" => paramOr("expires", "'expires' covered but no expires parameter")
        case _ =>
          if c.startsWith("@") then throw SigningBaseError(s"unsupported derived component: $component")
          headers.getOrElse(c, throw SigningBaseError(s"covered header $component not present"))
      lines += s"\"$c\": $value"
    }
    if mode == "rfc9421" then lines += s"\"@signature-params\": $spr"
    lines.mkString("\n")

  // ================= structured-field parse (port of parse.py) =================
  val LabelRe = "^\\s*([A-Za-z][A-Za-z0-9_-]*)\\s*=\\s*".r
  val CoveredRe = "^\\(\\s*(?:\"[^\"]*\"\\s*)*\\)".r

  def parseSignatureInput(header: String): Unit =
    if header == null then throw ParseError("header must be str")
    val hv = header.strip
    if hv.isEmpty then throw ParseError("empty header value")
    val rest = LabelRe.findPrefixMatchOf(hv) match
      case Some(m) => hv.substring(m.end)
      case None    => if hv.startsWith("(") then hv else throw ParseError("no label")
    if CoveredRe.findPrefixMatchOf(rest).isEmpty then throw ParseError("no covered list")

  def parseSignatureValue(header: String): Unit =
    if header == null then throw ParseError("header must be str")
    val hv = header.strip
    if hv.isEmpty then throw ParseError("empty Signature header value")
    val rest = LabelRe.findPrefixMatchOf(hv) match
      case Some(m) => hv.substring(m.end).strip
      case None    => if hv.startsWith(":") then hv else throw ParseError("no prefix")
    if !rest.startsWith(":") || !rest.endsWith(":") || rest.length < 2 then throw ParseError("not colon-wrapped")
    val sigB64 = rest.substring(1, rest.length - 1)
    val decoded = try Base64.getDecoder.decode(sigB64) catch case _: IllegalArgumentException => throw ParseError("bad base64")
    if Base64.getEncoder.encodeToString(decoded) != sigB64 then throw ParseError("non-canonical base64")

  // ================= Ed25519 key gate (port of keycheck.py) =================
  val P: BigInteger = BigInteger.TWO.pow(255).subtract(BigInteger.valueOf(19))
  val D: BigInteger = BigInteger.valueOf(-121665).multiply(BigInteger.valueOf(121666).modInverse(P)).mod(P)
  val SqrtM1: BigInteger = BigInteger.TWO.modPow(P.subtract(BigInteger.ONE).divide(BigInteger.valueOf(4)), P)
  val LOrder: BigInteger = BigInteger.TWO.pow(252).add(BigInteger("27742317777372353535851937790883648493"))

  def recoverX(y: BigInteger, sign: Int): BigInteger =
    if y.compareTo(P) >= 0 then return null
    val y2 = y.multiply(y).mod(P)
    val x2 = y2.subtract(BigInteger.ONE).mod(P)
      .multiply(D.multiply(y2).add(BigInteger.ONE).mod(P).modInverse(P)).mod(P)
    if x2.signum == 0 then return if sign == 1 then null else BigInteger.ZERO
    var x = x2.modPow(P.add(BigInteger.valueOf(3)).divide(BigInteger.valueOf(8)), P)
    if x.multiply(x).subtract(x2).mod(P).signum != 0 then x = x.multiply(SqrtM1).mod(P)
    if x.multiply(x).subtract(x2).mod(P).signum != 0 then return null
    if (if x.testBit(0) then sign == 0 else sign == 1) then x = P.subtract(x)
    x

  def decodePoint(pk: Array[Byte]): Array[BigInteger] =
    if pk.length != 32 then return null
    val le = Array.tabulate(32)(i => pk(31 - i))
    var y = new BigInteger(1, le)
    val sign = if y.testBit(255) then 1 else 0
    y = y.clearBit(255)
    val x = recoverX(y, sign)
    if x == null then null else Array(x, y, BigInteger.ONE, x.multiply(y).mod(P))

  def add(p: Array[BigInteger], q: Array[BigInteger]): Array[BigInteger] =
    val a = p(1).subtract(p(0)).multiply(q(1).subtract(q(0))).mod(P)
    val b = p(1).add(p(0)).multiply(q(1).add(q(0))).mod(P)
    val c = BigInteger.TWO.multiply(p(3)).multiply(q(3)).multiply(D).mod(P)
    val dd = BigInteger.TWO.multiply(p(2)).multiply(q(2)).mod(P)
    val e = b.subtract(a); val f = dd.subtract(c); val g = dd.add(c); val hh = b.add(a)
    Array(e.multiply(f).mod(P), g.multiply(hh).mod(P), f.multiply(g).mod(P), e.multiply(hh).mod(P))

  def mul8(p: Array[BigInteger]): Array[BigInteger] =
    val p2 = add(p, p); val p4 = add(p2, p2); add(p4, p4)

  def isIdentity(p: Array[BigInteger]): Boolean =
    p(0).mod(P).signum == 0 && p(1).subtract(p(2)).mod(P).signum == 0

  def isSmallOrder(pk: Array[Byte]): Boolean =
    val p = decodePoint(pk); p != null && isIdentity(mul8(p))

  def checkEd25519PublicKey(pk: Array[Byte]): Unit =
    val p = decodePoint(pk)
    if p == null then throw WeakKeyError("non-canonical/off-curve")
    if isIdentity(mul8(p)) then throw WeakKeyError("small-order")

  // ================= Ed25519 verify (port of verify.py) =================
  def ed25519Verify(msg: Array[Byte], sig: Array[Byte], pk: Array[Byte]): Boolean =
    if sig.length != 64 || pk.length != 32 then false
    else
      val gate = try { checkEd25519PublicKey(pk); true } catch case _: WeakKeyError => false
      if !gate then false
      else
        val signer = Ed25519Signer()
        signer.init(false, Ed25519PublicKeyParameters(pk, 0))
        signer.update(msg, 0, msg.length)
        signer.verifySignature(sig)

  // ================= ECDSA verify (port of algovoi-rfc9421-ecdsa) =================
  def ecdsaVerify(curve: String, msg: Array[Byte], sigRaw: Array[Byte], pub: Array[Byte], strict: Boolean): Boolean =
    try
      val fieldSize = if curve == "p256" then 32 else 48
      val stdName = if curve == "p256" then "secp256r1" else "secp384r1"
      val sigAlg = if curve == "p256" then "SHA256withECDSA" else "SHA384withECDSA"
      if sigRaw.length != 2 * fieldSize then return false
      val r = new BigInteger(1, sigRaw.slice(0, fieldSize))
      val s = new BigInteger(1, sigRaw.slice(fieldSize, 2 * fieldSize))
      val ap = AlgorithmParameters.getInstance("EC"); ap.init(ECGenParameterSpec(stdName))
      val spec = ap.getParameterSpec(classOf[ECParameterSpec]); val n = spec.getOrder
      if r.signum <= 0 || r.compareTo(n) >= 0 || s.signum <= 0 || s.compareTo(n) >= 0 then return false
      if strict && s.compareTo(n.shiftRight(1)) > 0 then return false
      if pub.length != 1 + 2 * fieldSize || (pub(0) & 0xff) != 0x04 then return false
      val x = new BigInteger(1, pub.slice(1, 1 + fieldSize))
      val yc = new BigInteger(1, pub.slice(1 + fieldSize, 1 + 2 * fieldSize))
      val pMod = spec.getCurve.getField.asInstanceOf[ECFieldFp].getP
      val lhs = yc.multiply(yc).mod(pMod)
      val rhs = x.multiply(x).multiply(x).add(spec.getCurve.getA.multiply(x)).add(spec.getCurve.getB).mod(pMod)
      if lhs != rhs then return false
      val key = KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(ECPoint(x, yc), spec))
      val v = ASN1EncodableVector(); v.add(ASN1Integer(r)); v.add(ASN1Integer(s))
      val der = DERSequence(v).getEncoded("DER")
      val verifier = Signature.getInstance(sigAlg); verifier.initVerify(key); verifier.update(msg)
      verifier.verify(der)
    catch case _: Exception => false

  // ================= RSA verify (rsa-pss-sha512 / rsa-v1_5-sha256) =================
  def rsaVerify(alg: String, base: Array[Byte], sig: Array[Byte], spki: Array[Byte]): Boolean =
    try
      val key = KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(spki))
      val s = alg match
        case "rsa-pss-sha512" =>
          val sg = Signature.getInstance("RSASSA-PSS")
          sg.setParameter(PSSParameterSpec("SHA-512", "MGF1", MGF1ParameterSpec.SHA512, 64, 1)); sg
        case "rsa-v1_5-sha256" => Signature.getInstance("SHA256withRSA")
        case _ => return false
      s.initVerify(key); s.update(base); s.verify(sig)
    catch case _: Exception => false

  def hex(s: String): Array[Byte] =
    val out = new Array[Byte](s.length / 2)
    var i = 0
    while i < s.length do { out(i / 2) = Integer.parseInt(s.substring(i, i + 2), 16).toByte; i += 2 }
    out

  def main(args: Array[String]): Unit =
    val path = if args.nonEmpty then args(0) else Option(System.getenv("ALGOVOI_NEGATIVE_V1")).getOrElse(DefaultPath)
    val corpus = ObjectMapper().readTree(new File(path))
    var total = 0; var matched = 0
    val fails = scala.collection.mutable.ArrayBuffer[String]()
    def rec(ok: Boolean, section: String, c: JsonNode): Unit =
      total += 1
      if ok then matched += 1
      else fails += s"[$section] " + (if c.hasNonNull("note") then c.get("note").asText() else "")

    corpus.get("signing_base").elements().asScala.foreach { c =>
      val want = new String(Base64.getDecoder.decode(c.get("signing_base_b64").asText()), StandardCharsets.UTF_8)
      val spr = if c.hasNonNull("signature_params_raw") then c.get("signature_params_raw").asText() else null
      val ok = try
        val built = buildSigningBase(c.get("in"), c.get("mode").asText(), spr)
        c.get("ok").asBoolean() && built == want
      catch case _: RuntimeException => !c.get("ok").asBoolean()
      rec(ok, "signing_base", c)
    }
    corpus.get("signature_input_parse").elements().asScala.foreach { c =>
      val p = try { parseSignatureInput(c.get("header").asText()); true } catch case _: ParseError => false
      rec(p == c.get("ok").asBoolean(), "sig_input_parse", c)
    }
    corpus.get("signature_value_parse").elements().asScala.foreach { c =>
      val p = try { parseSignatureValue(c.get("header").asText()); true } catch case _: ParseError => false
      rec(p == c.get("ok").asBoolean(), "sig_value_parse", c)
    }
    corpus.get("keygate").elements().asScala.foreach { c =>
      val raw = hex(c.get("pk_hex").asText())
      val rejected = try { checkEd25519PublicKey(raw); null } catch case _: WeakKeyError => "WeakKeyError"
      val want = if c.hasNonNull("rejected") then c.get("rejected").asText() else null
      rec(rejected == want && isSmallOrder(raw) == c.get("small_order").asBoolean(), "keygate", c)
    }
    corpus.get("ed25519_verify").elements().asScala.foreach { c =>
      val base = Base64.getDecoder.decode(c.get("signing_base_b64").asText())
      val valid = ed25519Verify(base, hex(c.get("sig_hex").asText()), hex(c.get("pk_hex").asText()))
      rec(valid == c.get("expect_valid").asBoolean(), "ed25519_verify", c)
    }
    corpus.get("ecdsa_verify").elements().asScala.foreach { c =>
      val strict = c.hasNonNull("strict_low_s") && c.get("strict_low_s").asBoolean()
      val valid = ecdsaVerify(c.get("curve").asText(), hex(c.get("msg_hex").asText()),
        hex(c.get("sig_raw_hex").asText()), hex(c.get("pub_uncompressed_hex").asText()), strict)
      rec(valid == c.get("expect_valid").asBoolean(), "ecdsa_verify", c)
    }

    if corpus.has("rsa_verify") then corpus.get("rsa_verify").elements().asScala.foreach { c =>
      val base = Base64.getDecoder.decode(c.get("signing_base_b64").asText())
      val valid = rsaVerify(c.get("alg").asText(), base, hex(c.get("sig_hex").asText()), hex(c.get("pub_spki_hex").asText()))
      rec(valid == c.get("expect_valid").asBoolean(), "rsa_verify", c)
    }

    fails.foreach(f => println(s"FAIL  $f"))
    println(s"\nscala: $matched/$total cases matched")
    System.exit(if matched == total then 0 else 1)
